File: src/main/java/com/ottofernan/librarycrud/controllers/VisitorController.java
package com.ottofernan.librarycrud.controllers;

import com.ottofernan.librarycrud.domain.dtos.BookDTO;
import com.ottofernan.librarycrud.domain.models.Visitor;
import com.ottofernan.librarycrud.domain.models.VisitorBook;
import com.ottofernan.librarycrud.services.restbook.RestBookService;
import com.ottofernan.librarycrud.services.visitor.VisitorService;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;
import org.springframework.web.servlet.support.RequestContextUtils;

import javax.servlet.http.HttpServletRequest;
import java.util.Map;

import static com.ottofernan.librarycrud.domain.dtos.BookDTOKt.toModel;

@Controller
@RequestMapping("/visitors")
public class VisitorController {

    private final VisitorService visitorService;
    private final RestBookService restBookService;

    public VisitorController(VisitorService visitorService, RestBookService restBookService) {
        this.visitorService = visitorService;
        this.restBookService = restBookService;
    }

    @GetMapping("/signIn")
    public String signInPage (@ModelAttribute("visitor") Visitor visitor, Model model){

        if(visitor == null) {
            model.addAttribute("visitor", new Visitor());
        } else if(visitor.getFirstName() != null && visitor.getLastName() != null){
            visitorService.save(visitor);
            return "index";
        }

        return "/visitors/signIn";
    }
    

    @GetMapping("/auth")
    public String auth(HttpServletRequest request, RedirectAttributes redirect, Model model){

        Map<String, ?> inputFlashMap = RequestContextUtils.getInputFlashMap(request);
        if(inputFlashMap != null) {
            inputFlashMap.forEach(redirect::addFlashAttribute);
            VisitorBook visitor = (VisitorBook) inputFlashMap.get("visitor");
            Visitor correctVisitor = visitorService.findByFirstName(visitor
                    .getVisitor().getFirstName());

            if (Visitor.isPasswordCorrect(visitor.getVisitor(), correctVisitor)) {
                String nextPage = (String) inputFlashMap.get("successNextPage");
                return "redirect:" + nextPage;
            } else if (correctVisitor == null) {
                model.addAttribute("visitor", new Visitor());
                return "/visitors/signIn";
            } else {
                String nextPage = (String) inputFlashMap.get("failNextPage");
                return "redirect:" + nextPage;
            }
        }

        return "index";
    }

    @GetMapping("/rentBook")
    public String rentBook(@ModelAttribute VisitorBook visitorBook, RedirectAttributes redirectAttributes){
        Visitor visitor = visitorBook.getVisitor();
        System.out.println(visitorBook.getBook().getId());

        if(Visitor.isNotValid(visitor))
            return "redirect:http://localhost:8080/books/rentBook";

        redirectAttributes.addFlashAttribute("successNextPage", "/visitors/rentAuthTrue");
        redirectAttributes.addFlashAttribute("failNextPage", "/visitors/rentAuthFalse");
        redirectAttributes.addFlashAttribute("visitor", visitorBook);
        return "redirect:/visitors/auth";

    }

    @GetMapping("/rentAuthTrue")
    public String executeRentBook(HttpServletRequest request){

        Map<String, ?> inputFlashMap = RequestContextUtils.getInputFlashMap(request);
        if(inputFlashMap == null)
            return "error_";

        VisitorBook visitorBook = (VisitorBook) inputFlashMap.get("visitor");
        BookDTO book = restBookService.findById(visitorBook.getBook().getId());
        System.out.println(visitorBook);
        Visitor visitor = visitorService.findByFirstName(visitorBook.getVisitor().getFirstName());
        if(toModel(book).rent(visitor)) {
            visitorService.save(visitor);
            restBookService.update(book);
            return "books/rentSuccessfully";
        } else {
            return "books/alreadyHaveBook";
        }
    }

    @GetMapping("/rentAuthFalse")
    public String failOnRent(){
        return "visitors/wrongPassword";
    }

    @GetMapping("/returnAuthTrue")
    public String returnSucessfully(HttpServletRequest request, Model model){
        Map<String, ?> input = RequestContextUtils.getInputFlashMap(request);
        if(input != null){
            VisitorBook visitorBook = (VisitorBook) input.get("visitor");
            Visitor checkedVisitor = visitorService.findByFirstName(visitorBook.getVisitor().getFirstName());
            visitorBook.setVisitor(checkedVisitor);
            model.addAttribute("visitor", visitorBook);
            return "/books/listVisitorsBooks";
        }
        return "error_";
    }

    @GetMapping("/returnAuthFail")
    public String failOnReturn(){
        return "visitors/wrongPassword";
    }

    @GetMapping("/executeReturn")
    public String executeReturn(@ModelAttribute("visitor") VisitorBook visitorBook, RedirectAttributes redirect){

        Visitor checkedVisitor = visitorService.findById(visitorBook.getVisitor().getId());
        BookDTO checkedBook = restBookService.findById(visitorBook.getBook().getId());

        if(checkedBook == null || checkedVisitor == null){
            redirect.addFlashAttribute("visitor", visitorBook);
            return "redirect:/visitors/returnAuthTrue";
        }

        if(toModel(checkedBook).returnBook(checkedVisitor)) {
            visitorService.save(checkedVisitor);
            restBookService.update(checkedBook);
        } else {
            redirect.addFlashAttribute("visitor", visitorBook);
            return "redirect:/visitors/returnAuthTrue";
        }

        return "redirect:/index";
    }

}

File: src/main/java/com/ottofernan/librarycrud/controllers/VisitorController.kt
package com.ottofernan.librarycrud.controllers

import com.ottofernan.librarycrud.domain.dtos.VisitorDTO
import com.ottofernan.librarycrud.domain.dtos.toModel
import com.ottofernan.librarycrud.domain.models.Visitor
import com.ottofernan.librarycrud.domain.models.VisitorBook
import com.ottofernan.librarycrud.services.restbook.RestBookService
import com.ottofernan.librarycrud.services.visitor.VisitorService
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.servlet.mvc.support.RedirectAttributes
import org.springframework.web.servlet.support.RequestContextUtils
import javax.servlet.http.HttpServletRequest

@Controller
@RequestMapping("/visitors")
class VisitorController(
        private val visitorService: VisitorService,
        private val restBookService: RestBookService
) {

    @GetMapping("/signIn")
    fun signInPage(@ModelAttribute("visitor") visitor: VisitorDTO?, model: Model): String {
        if(visitor == null) {
            model.addAttribute("visitor", VisitorDTO())
            return "/visitors/signIn"
        }

        visitorService.save(visitor)
        return "redirect:/"

    }

    @GetMapping("/auth")
    fun auth (request: HttpServletRequest, redirect: RedirectAttributes, model: Model): String {

        val inputFlashMap = RequestContextUtils.getInputFlashMap(request) ?: return "redirect:/"
        inputFlashMap.forEach { (s: String?, o: Any?) -> redirect.addFlashAttribute(s, o) }

        val visitor = inputFlashMap["visitor"] as VisitorBook
        val correctVisitor = visitorService.findByFirstName(visitor.visitor.firstName)
        if(correctVisitor == null){
            model.addAttribute("visitor", VisitorDTO())
            return "visitors/signIn"
        }

        val nextPage: String
        nextPage = if (Visitor.isPasswordCorrect(visitor.visitor, correctVisitor))
            inputFlashMap["successNextPage"] as String
        else inputFlashMap["failNextPage"] as String

        return "redirect:$nextPage"
    }

    @GetMapping("/rentBook")
    fun rentBook(@ModelAttribute visitorBook: VisitorBook, redirect: RedirectAttributes): String {
        val visitor = visitorBook.visitor
        if(!Visitor.isValid(visitor)) return "redirect:http://localhost:8080/books/rentBook"

        redirect.addFlashAttribute("successNextPage", "/visitors/rentAuthTrue")
        redirect.addFlashAttribute("failNextPage", "/visitors/rentAuthFalse")
        redirect.addFlashAttribute("visitor", visitorBook)
        return "redirect:/visitors/auth"
    }

    @GetMapping("/rentAuthTrue")
    fun executeRentBook(request: HttpServletRequest): String {
        val inputFlashMap = RequestContextUtils.getInputFlashMap(request) ?: return "error_"

        val visitorBook = inputFlashMap["visitor"] as VisitorBook
        val bookDTO = restBookService.findById(visitorBook.book.id)
        val visitor = visitorService.findByFirstName((visitorBook.visitor.firstName))
        val book = toModel(bookDTO)

        if(!book.rent(visitor)) return "books/alreadyHaveBook"

        visitorService.save(visitor)
        restBookService.update(book.toDto())
        return "books/rentSuccessfully"
    }

    @GetMapping("/rentAuthFalse")
    fun failOnRent(): String =
        "visitors/wrongPassword"

    @GetMapping("/returnAuthTrue")
    fun returnSuccessfully (request: HttpServletRequest, model: Model): String{
        val input = RequestContextUtils.getInputFlashMap(request) ?: return "error_"
        val visitorBook = input["visitor"] as VisitorBook
        val checkedVisitor = visitorService.findByFirstName(visitorBook.visitor.firstName)
        visitorBook.visitor = checkedVisitor
        model.addAttribute("visitor", visitorBook)
        return "/books/listVisitorsBooks"

    }

    @GetMapping("/returnAuthFail")
    fun failOnReturn(): String =
            "visitors/wrongPassword"

    @GetMapping("/executeReturn")
    fun executeReturn(@ModelAttribute("visitor") visitorBook: VisitorBook, redirect: RedirectAttributes): String {
        val checkedVisitor = visitorService.findById(visitorBook.visitor.id)
        val checkedBookDTO = restBookService.findById(visitorBook.book.id)
        val checkedBook = toModel(checkedBookDTO)

        if(checkedVisitor == null) {
            redirect.addFlashAttribute("visitor", visitorBook)
            return "redirect:/visitors/returnAuthTrue"
        }

        return if(checkedBook.returnBook(checkedVisitor)) {
            visitorService.save(checkedVisitor)
            restBookService.update(checkedBook.toDto())
            "redirect:/"
        } else {
            redirect.addFlashAttribute("visitor", visitorBook)
            "redirect:/visitors/returnAuthTrue"
        }
    }
}